src/slice-chop.c

```diff
@@ -380,6 +380,185 @@ static SEXP chop_fallback_shaped(SEXP x, SEXP indices, struct vctrs_chop_info in
   return info.out;
 }
 
+// -----------------------------------------------------------------------------
+
+// From type.c
+SEXP vctrs_type_common_impl(SEXP dots, SEXP ptype);
+
+// From slice-assign.c
+SEXP vec_assign_impl(SEXP proxy, SEXP index, SEXP value, bool clone);
+
+static SEXP vec_unchop(SEXP x, SEXP indices, SEXP ptype, const struct name_repair_opts* name_repair);
+
+// [[ register() ]]
+SEXP vctrs_unchop(SEXP x, SEXP indices, SEXP ptype, SEXP name_repair) {
+  struct name_repair_opts name_repair_opts = new_name_repair_opts(name_repair, false);
+  PROTECT_NAME_REPAIR_OPTS(&name_repair_opts);
+
+  SEXP out = vec_unchop(x, indices, ptype, &name_repair_opts);
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP vec_unchop_indices(SEXP out,
+                               SEXP* p_out_names,
+                               PROTECT_INDEX out_pi,
+                               SEXP x,
+                               R_len_t x_size,
+                               bool is_shaped,
+                               bool has_inner_names,
+                               SEXP indices);
+
+static SEXP vec_unchop_sequentially(SEXP out,
+                                    SEXP* p_out_names,
+                                    PROTECT_INDEX out_pi,
+                                    SEXP x,
+                                    R_len_t x_size,
+                                    bool is_shaped,
+                                    bool has_inner_names);
+
+static SEXP vec_unchop(SEXP x, SEXP indices, SEXP ptype, const struct name_repair_opts* name_repair) {
+  if (TYPEOF(x) != VECSXP) {
+    Rf_errorcall(R_NilValue, "`x` must be a list");
+  }
+
+  R_len_t x_size = vec_size(x);
+
+  R_len_t out_size = 0;
```

Comment:
Could there be a way to give that information upfront, typically for the use case I have in mind in `mutate()` we "already know".

Reply:
I'm not sure it would end up being useful in the end. You still have to recycle each element of `x` to the size of the corresponding index element. So you have to compute the index sizes anyways

https://github.com/r-lib/vctrs/pull/834/files#diff-04334dad60bfe91381c3ac91164f5429R461-R472